src/userobjects/OpenMCDomainFilterEditor.C

```diff
@@ -0,0 +1,202 @@
+#include "openmc/tallies/filter_mesh.h"
+
+registerMooseObject("CardinalApp", OpenMCDomainFilterEditor);
+
+InputParameters
+OpenMCDomainFilterEditor::validParams()
+{
+  InputParameters params = OpenMCUserObject::validParams();
+  params.addParam<bool>("create_filter", false, "Whether to create the tally if it doesn't exist");
+  params.addRequiredParam<int32_t>("filter_id", "The ID of the filter to modify");
+  params.addParam<MooseEnum>("filter_type", getFilterTypeEnum(), "The type of filter to create");
+  params.addRequiredParam<std::vector<std::string>>("bins", "The bins to modify in the filter");
+  params.declareControllable("bins");
+  params.addClassDescription("A UserObject for creating and managing OpenMC domain tally filters");
+  return params;
+}
+
+OpenMCDomainFilterEditor::OpenMCDomainFilterEditor(const InputParameters & parameters)
+  : OpenMCUserObject(parameters),
+    _create_filter(getParam<bool>("create_filter")),
+    _filter_id(getParam<int32_t>("filter_id")),
+    _filter_type(getParam<MooseEnum>("filter_type").getEnum<OpenMCFilterType>())
+{
+
+  if (!parameters.isParamSetByUser("filter_type"))
+  {
+    mooseError(longName() + ": filter_type must be specified");
+  }
+
+  this->initialize();
+}
+
+void
+OpenMCDomainFilterEditor::initialize()
+{
+  bool filter_exists = this->filterExists();
+
+  // if create_filter is set to true, but the filter already exists, display a warning
+  if (_create_filter && filter_exists)
+  {
+    mooseWarning(longName() + ": Filter " + std::to_string(_filter_id) +
+                 " already exists in the OpenMC XML model");
+  }
+
+  if (!_create_filter && !filter_exists)
+    paramError("filter_id", "Filter " + std::to_string(_filter_id) + " does not exist and create_filter is false");
+
+  // if the filter doesn't exist at this point and no other errors have been raised,
+  // create the filter
+  if (!filter_exists)
+    openmc::Filter::create(filterTypeEnumToString(_filter_type), _filter_id);
+
+  // at this point the filter exists and the filter type is set, check that the type is valid
+  checkFilterTypeMatch();
+}
+
+void
+OpenMCDomainFilterEditor::checkFilterTypeMatch() const
+{
+  // check this UO's filter type against the one in the OpenMC model
+  std::string existing_type_str = openmc::model::tally_filters[filterIndex()]->type_str();
+  OpenMCFilterType existing_type = stringToFilterTypeEnum(existing_type_str);
+  _console << "CHECKING FILTER TYPE..." << std::endl;
+
+  if (existing_type != _filter_type)
+    paramError("filter_id", "An existing filter, Filter " + std::to_string(_filter_id) + ", is of type \"" + existing_type_str + "\" and cannot be changed to type \"" + filterTypeEnumToString(_filter_type) + "\"");
+}
+
+bool
+OpenMCDomainFilterEditor::filterExists() const
+{
+  return openmc::model::filter_map.find(_filter_id) != openmc::model::filter_map.end();
+}
+
+int32_t
+OpenMCDomainFilterEditor::filterIndex() const
+{
+  if (!filterExists())
```

Comment:
I don't think this error message is tested anywhere. @pshriwise would you please add a test to cover it?

Reply:
Because the `initialize()` method is now called right away we shouldn't need that check anymore. Removing.